File: santa/src/santa.c
#include <stdio.h>
#include <stdbool.h>
#include <pthread.h>
#include <semaphore.h>
#define P(sem)  (sem_wait(&(sem)))  /* uses P and V for the wait and ... */
#define V(sem)  (sem_post(&(sem)))  /* ... signal semaphore operations */

sem_t wakeup, wakeupReindeer, wakeupElves;
sem_t harness, harnessDone;
sem_t pull, pullDone;
sem_t enter, enterDone;
sem_t consult, consultDone;
sem_t reindeerBack, reindeerBackDone;
sem_t reindeerHarness, reindeerHarnessDone;
sem_t reindeerPull, reindeerPullDone;
sem_t elfPuzzled, elfPuzzledDone;
sem_t elfEnter, elfEnterDone;
sem_t elfConsult, elfConsultDone;

bool b;

void *Santa(void *arg) {
    for (int t = 0; t < 100000; t++) { // Sleeping
        // printf("Santa sleeping\n");
        P(wakeup); // woken up by Sleigh or Shop
        if (b) { // Delivering
            b = false; V(wakeupReindeer);
            // printf("Santa harnessing\n");
            P(harness); V(harnessDone);
            // printf("Santa riding\n");
            P(pull); V(pullDone);
        } else { // Helping
            V(wakeupElves);
            for (int i = 0; i < 3; i++) {
                // printf("Santa welcoming\n");
                P(enter); V(enterDone);
                // printf("Santa explaining\n");
                P(consult); V(consultDone);
            }
        }
    }
}
void *Sleigh(void *arg) {
    for (;;) {
        for (int i = 0; i < 9; i++) V(reindeerBack);
        for (int i = 0; i < 9; i++) P(reindeerBackDone);
        // printf("9 reindeer back\n");
        b = true; V(wakeup); P(wakeupReindeer);
        // printf("9 reindeer stating to harness\n");
        for (int i = 0; i < 9; i++) V(reindeerHarness);
        for (int i = 0; i < 9; i++) P(reindeerHarnessDone);
        // printf("9 reindeer done harnessing\n");
        V(harness); P(harnessDone);
        // printf("9 reindeer stating to pull\n");
        for (int i = 0; i < 9; i ++) V(reindeerPull);
        for (int i = 0; i < 9; i ++) P(reindeerPullDone);
        // printf("9 reindeer done pulling\n");
        V(pull); P(pullDone);
        // printf("Sleigh starting over\n");
    }
}
void *Reindeer(void *arg) {
    for (int t = 0; t < 20000; t++) {
        P(reindeerBack); V(reindeerBackDone);
        P(reindeerHarness); V(reindeerHarnessDone);
        P(reindeerPull); V(reindeerPullDone);
    }
}
void *Shop(void *arg) {
    for (;;) {
        for (int i = 0; i < 3; i++) V(elfPuzzled);
        for (int i = 0; i < 3; i++) P(elfPuzzledDone);
        // printf("3 elves puzzled\n");
        V(wakeup); P(wakeupElves);
        for (int i = 0; i < 3; i++) {
            V(elfEnter); P(elfEnterDone);
            // printf("1 elf entered\n");
            V(enter); P(enterDone);
            V(elfConsult); P(elfConsultDone);
            // printf("1 elf consulting\n");
            V(consult); P(consultDone);
        }
    }
}
void *Elf(void *arg) {
    for (;;) {
        P(elfPuzzled); V(elfPuzzledDone);
        P(elfEnter); V(elfEnterDone);
        P(elfConsult); V(elfConsultDone);
    }
}
void main() {
    sem_init(&wakeup, 0, 0); sem_init(&wakeupReindeer, 0, 0); sem_init(&wakeupElves, 0, 0);
    sem_init(&harness, 0, 0); sem_init(&harnessDone, 0, 0);
    sem_init(&pull, 0, 0); sem_init(&pullDone, 0, 0);
    sem_init(&enter, 0, 0); sem_init(&enterDone, 0, 0);
    sem_init(&consult, 0, 0); sem_init(&consultDone, 0, 0);
    sem_init(&reindeerBack, 0, 0); sem_init(&reindeerBackDone, 0, 0);
    sem_init(&reindeerHarness, 0, 0); sem_init(&reindeerHarnessDone, 0, 0);
    sem_init(&reindeerPull, 0, 0); sem_init(&reindeerPullDone, 0, 0);
    sem_init(&elfPuzzled, 0, 0); sem_init(&elfPuzzledDone, 0, 0);
    sem_init(&elfEnter, 0, 0); sem_init(&elfEnterDone, 0, 0);
    sem_init(&elfConsult, 0, 0); sem_init(&elfConsultDone, 0, 0);
    pthread_t tid;
    for (int i = 0; i < 9; i++) pthread_create(&tid, NULL, Reindeer, NULL);
    for (int i = 0; i < 20; i++) pthread_create(&tid, NULL, Elf, NULL);
    pthread_create(&tid, NULL, Sleigh, NULL); pthread_create(&tid, NULL, Shop, NULL);
    pthread_create(&tid, NULL, Santa, NULL); pthread_join(tid, NULL);
}

File: santa/src/santa.java
enum R {Relaxing, Back, Harnessing, Harnessed, Pulling, Done}
enum E {Working, Puzzled, Entering, Entered, Consulting, Enlightened}
enum Task {deliver, help}
class SantasShop {
    int rc = 9, ec = 3; // reindeer count, elf count
    R rs = R.Relaxing;  // state of reindeer
    E es = E.Working;   // state of elves

    synchronized void back() /* called by reindeer */ {
        while (rs != R.Relaxing) try {wait();} catch (Exception x) {}
        rc -= 1; if (rc == 0) {rs = R.Back; rc = 9;} notifyAll();
    }
    synchronized void harness() /* called by reindeer */ {
        while (rs != R.Harnessing) try {wait();} catch (Exception x) {}
        rc -= 1; if (rc == 0) {rs = R.Harnessed; rc = 9;} notifyAll();
    }
    synchronized void pull() /* called by reindeer */ {
        while (rs != R.Pulling) try {wait();} catch (Exception x) {}
        rc -= 1; if (rc == 0) {rs = R.Done; rc = 9;} notifyAll();
    }

    synchronized void puzzled() /* called by elves */ {
        while (es != E.Working) try {wait();} catch (Exception x) {}
        ec -= 1; if (ec == 0) {es = E.Puzzled; ec = 3;} notifyAll();
    }
    synchronized void enter() /* called by elves */ {
        while (es != E.Entering) try {wait();} catch (Exception x) {}
        es = E.Entered; notifyAll();
    }
    synchronized void consult() /* called by elves */ {
        while (es != E.Consulting) try {wait();} catch (Exception x) {}
        es = E.Enlightened; notifyAll();
    }

    synchronized Task wakeup() /* called by Santa */ {
        while (rs != R.Back && es != E.Puzzled) try {wait();} catch (Exception x) {}
        if (rs == R.Back) {rs = R. Harnessing; notifyAll(); return Task.deliver;}
        else {es = E.Entering; notifyAll(); return Task.help;}
    }
    synchronized void hitch() /* called by Santa */ {
        while (rs != R.Harnessed) try {wait();} catch (Exception x) {}
        rs = R.Pulling; notifyAll();
    }
    synchronized void ride() /* called by Santa */ {
        while (rs != R.Done) try {wait();} catch (Exception x) {}
        rs = R.Relaxing; notifyAll();
    }
    synchronized void welcome() /* called by Santa */ {
        while (es != E.Entered) try {wait();} catch (Exception x) {}
        es = E.Consulting; notifyAll();
    }
    synchronized void explain() /* called by Santa */ {
        while (es != E.Enlightened) try {wait();} catch (Exception x) {}
        ec -= 1; if (ec == 0) {es = E.Working; ec = 3;} else es = E.Entering;
        notifyAll();
    }

    public static void main(String[] args) {
        SantasShop shop = new SantasShop();
        new Santa(shop).start();
        for (int i = 0; i < 9; i++) new Reindeer(shop).start();
        for (int i = 0; i < 20; i++) {Thread e = new Elf(shop, i); e.setDaemon(true); e.start();}
    }
}
class Santa extends Thread {
    SantasShop shop;
    Santa(SantasShop ss) {shop = ss;}
    public void run() {
        for (int t = 0; t < 10000; t++) {
            Task task = shop.wakeup();
            if (task == Task.deliver) {
                shop.hitch(); shop.ride();
            } else {
                for (int i = 0; i < 3; i++) {shop.welcome(); shop.explain();}
            }
        }
    }
}
class Reindeer extends Thread {
    SantasShop shop;
    Reindeer(SantasShop ss) {shop = ss;}
    public void run() {
        for (int t = 0; t < 2000; t++) {shop.back(); shop.harness(); shop.pull();}
    }
}
class Elf extends Thread {
    SantasShop shop; int num;
    Elf(SantasShop ss, int n) {shop = ss; num = n;}
    public void run() {
        for (;;) {shop.puzzled(); shop.enter(); shop.consult();}
    }
}
